### PR TITLE
RootKeyProvider implementation update to enable its use in middleware

### PR DESCRIPTION
**Pull request by Tipnos**
This PR is the continuation of https://github.com/biscuit-auth/biscuit-rust/pull/165 and is related to https://github.com/biscuit-auth/biscuit-actix-middleware/pull/8.

It contains necessary changes to allow use of `RootKeyProvider` in middlewares:
- `Box`, `Rc`, `Arc` implementations (actix uses `Rc`, the rest can be useful for other web frameworks)
- Change `RootKeyProvider` functions argument by `&RootKeyProvider` to  avoid cloning `PublicKey` list foreach call https://github.com/Tipnos/biscuit-rust/commit/fca6c48a722de5941935b11a41bb3cfa40f0efcb

### CODE REVIEW

**Comment by divarvel** [comment]
This is technically a breaking change, since the key provider is now taken by reference. I think it's better now, but that might warrant a major version bump.

File: biscuit-auth/src/format/mod.rs
Comment (Geal) [inline biscuit-auth/src/format/mod.rs:43]: would this be posible with a `AsRef` to avoid an API breaking change?

Comment (Tipnos) [reply biscuit-auth/src/format/mod.rs:43]: No problem! 

Sorry for being picky, would you agree to use `Borrow` instead of `AsRef`? Indeed, `Borrow` has more garantee like `In particular Eq, Ord and Hash must be equivalent for borrowed and owned values: x.borrow() == y.borrow() should give the same result as x == y`

File: biscuit-auth/src/lib.rs
Comment (divarvel) [inline biscuit-auth/src/lib.rs:62]: it's good to know it works well in both cases, but maybe we should harmonize the examples?

imo passing it by reference should be the default in examples. If we want to make sure it works both ways (and we should!), ensuring it in tests would be good and would not affect documentation.

Comment (Tipnos) [reply biscuit-auth/src/lib.rs:62]: I put both to show all possibilities as you were concerned by API breaking changes. 

But IMO I find reference only examples better!

I'm waiting for @Geal approval to make final modifications.

**Review by Geal — APPROVED** [review]

File: biscuit-auth/src/lib.rs
Comment (Geal) [reply biscuit-auth/src/lib.rs:62]: references should work better for the examples yes

**Comment by Tipnos** [comment]
Bad news!

It looks like we can't allow passing `&RootKeyProvider` as function argument without introducing breaking change.

As we saw earlier we have to introduce `Borrow` trait to be able to pass by value and reference. 

But thanks to a nice @divarvel final review catch, closure impl was no more tested. Indeed, I removed the test case by misunderstanding of its use.

To pass the test the following implementation is required:
```rust 
impl<F: Fn(Option<u32>) -> Result<crypto::PublicKey, error::Format>> Borrow<dyn RootKeyProvider>
    for F
{
    fn borrow(&self) -> &(dyn RootKeyProvider + 'static) {
        self
    }
} 
```

[But rust does not allow it](https://rust-lang.github.io/rfcs/2451-re-rebalancing-coherence.html).

An other alternative could be:
```rust
impl Borrow<dyn RootKeyProvider>
    for dyn Fn(Option<u32>) -> Result<crypto::PublicKey, error::Format>
{
    fn borrow(&self) -> &(dyn RootKeyProvider + 'static) {
        self
    }
} 
```

But cast between traits is not allowed.

Final alternative:
```rust
impl Borrow<dyn RootKeyProvider> for fn(Option<u32>) -> Result<crypto::PublicKey, error::Format> {
    fn borrow(&self) -> &(dyn RootKeyProvider + 'static) {
        self
    }
}
```

This implementation compiles but as the closure has an argument (capture its env) it can't be cast as a function pointer.

After discussion with @divarvel, we think the best solution is to put the subject aside until next major release. 

Concerning https://github.com/biscuit-auth/biscuit-actix-middleware/pull/8 we'll clone `Vec<PublicKey>` on every incoming calls until next major release.

PS: if performance issues about cloning occure before next major update, we still have the possibility to temporarily add a new function that take a reference.

**Comment by Tipnos** [comment]
See https://github.com/biscuit-auth/biscuit-rust/pull/168 for final merged implementation.

**Pull request closed without merging**
